### DISCUSSION TITLE
How to prerender Single Page Application (SPA) in Vercel?

### DISCUSSION THREAD

**Discussion by kresnasatya** [vercel/community#3306] [Help] [answered]
Hi everyone, I make a site called Brave Frontier Wiki Unofficial. This wiki shows list of omniunits and DBB of Brave Frontier Games.
The site built with Vercel monorepo term and the site based on Single Page Application. I'm using vanilla JavaScript instead of frameworks.

Here's the repository code: https://github.com/satyakresna/xbravefrontier and the link of site: https://bravefrontier.kresna.me/.

**I have a problem with opengraph issue. When I check the page of my site: https://bravefrontier.kresna.me/omniunits/Blast_Brigadier_Buck in OpenGraph.xyz site: https://www.opengraph.xyz/url/https:%2F%2Fbravefrontier.kresna.me%2Fomniunits%2FBlast_Brigadier_Buck/ the thumbnail and name of character don't show. In other words it's not dynamic.**

![Screenshot 2021-05-14 12 07 21](https://user-images.githubusercontent.com/11099186/118220122-78905980-b4ad-11eb-9910-96cfef8a38a4.png)

**I thought why the thumbnail and name of character don't show because of the single page application it self doesn't pre rendered.** I check in Vercel article about [Introducing Serverless Pre-Rendering (SPR)](https://vercel.com/blog/serverless-pre-rendering) but I'm not sure if that's the solution I'm looking for.

Have you been face the same case and solved it? If yes, could you please you shared about how to solved it? Thank you.

**Comment by leerob** [comment, marked as answer]
https://vercel.com/support/articles/how-can-i-prerender-my-application-on-vercel

For those who found this from Google 😄 (Sorry for the late response)

**Reply by kresnasatya** [reply]
Thanks Lee Rob. I think I should migrate my project into frameworks like NextJS or Sveltekit. If I keep in Vanilla way I need more extra effort to build it in Static Site Generator or Server Side Rendering. 😁

**Reply by movahhedi** [reply]
How can I SSR with vanilla js? I really love my codebase, but SEO is always an issue for me. Is it possible to use prerender.io on Vercel?

**Reply by leerob** [reply]
I would recommend using a tool that natively supports SSR, whether that's Next.js (React), Nuxt (Vue), SvelteKit (Svelte), or similar.

**Comment by dolvexsanit23** [comment]


**Reply by movahhedi** [reply]
What if I don't use React or Vue (Next.js or Nuxt.js)? Pure JS SPA (e.g. with Babel).
